src/osdc/ObjectCacher.cc

```diff
@@ -580,11 +581,18 @@ void ObjectCacher::Object::truncate(loff_t s)
 
     // remove bh entirely
     ceph_assert(bh->start() >= s);
-    ceph_assert(bh->waitfor_read.empty());
+    for ([[maybe_unused]] auto& [off, ctxs] : bh->waitfor_read) {
+      waiting_for_read.splice(waiting_for_read.end(), ctxs);
+    }
```

Comment:
why not just call
```c++
for ([[maybe_unused]] auto& [off, ctxs] : bh->waitfor_read) {
  for (auto waiter : ctxs) {
    waiter->complete(0);
  }
}
```

also. "0" denotes that the read is still in progress. not sure if we should pass "0" to the waiters in this case.

Reply:
I'm following the style in other parts of the code here because I think the concern is that the waiter retry may push on to the list.

0 also looks to be the way to signal retry here: https://github.com/ceph/ceph/blob/6bd8fa86796ed3e1a40e13491340c9f1b4a0f18f/src/osdc/ObjectCacher.cc#L1802